network/network-address/src/lib.rs

```diff
@@ -643,6 +685,124 @@ impl Arbitrary for DnsName {
     }
 }
 
+/////////////
+// Parsing //
+/////////////
+
+/// parse the `&[Protocol]` into the `"/memory/<port>"` prefix and unparsed
+/// `&[Protocol]` suffix.
+pub fn parse_memory(protos: &[Protocol]) -> Option<(u16, &[Protocol])> {
+    protos
+        .split_first()
+        .and_then(|(first, suffix)| match first {
+            Protocol::Memory(port) => Some((*port, suffix)),
+            _ => None,
+        })
+}
+
+/// parse the `&[Protocol]` into the `"/ip4/<addr>/tcp/<port>"` or
+/// `"/ip6/<addr>/tcp/<port>"` prefix and unparsed `&[Protocol]` suffix.
+pub fn parse_ip_tcp(protos: &[Protocol]) -> Option<((IpAddr, u16), &[Protocol])> {
+    use Protocol::*;
+
+    if protos.len() < 2 {
+        return None;
+    }
+
+    let (prefix, suffix) = protos.split_at(2);
+    match prefix {
+        [Ip4(ip), Tcp(port)] => Some(((IpAddr::V4(*ip), *port), suffix)),
+        [Ip6(ip), Tcp(port)] => Some(((IpAddr::V6(*ip), *port), suffix)),
+        _ => None,
+    }
+}
+
+/// parse the `&[Protocol]` into the `"/dns/<domain>/tcp/<port>"`,
+/// `"/dns4/<domain>/tcp/<port>"`, or `"/dns6/<domain>/tcp/<port>"` prefix and
+/// unparsed `&[Protocol]` suffix.
+pub fn parse_dns_tcp(protos: &[Protocol]) -> Option<((&DnsName, u16), &[Protocol])> {
+    use Protocol::*;
+
+    if protos.len() < 2 {
+        return None;
+    }
+
+    let (prefix, suffix) = protos.split_at(2);
+    match prefix {
+        [Dns(name), Tcp(port)] => Some(((name, *port), suffix)),
+        [Dns4(name), Tcp(port)] => Some(((name, *port), suffix)),
+        [Dns6(name), Tcp(port)] => Some(((name, *port), suffix)),
```

Comment:
Was looking at where this was used, we just pass in name into tokio. This means tokio might connect to either 4 or 6 but not necessarily what we specified. I recommend we consider leveraging Rust's ToSockAdder and select the right name if 4 or 6 is specified. If it is just Dns, we can delegate parsing to tokio. Thoughts?

Reply:
I think we can just use [`tokio::net::lookup_host`](https://docs.rs/tokio/0.2.21/tokio/net/fn.lookup_host.html) directly and filter the results, e.g., `Dns4` will only use `SocketAddrV4`, `Dns6` will only use `SocketAddrV6`, and `Dns` will use either.